File: src/org/simbrain/plot/rasterchart/RasterPlotPanel.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.plot.rasterchart;

import org.jfree.chart.ChartFactory;
import org.jfree.chart.ChartPanel;
import org.jfree.chart.JFreeChart;
import org.jfree.chart.plot.PlotOrientation;
import org.jfree.chart.plot.XYPlot;
import org.jfree.chart.renderer.xy.XYItemRenderer;
import org.jfree.data.xy.XYSeriesCollection;
import org.simbrain.util.propertyeditor.gui.ReflectivePropertyEditor;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Rectangle2D;

/**
 * Display a raster plot. This component can be used independently of the
 * raster plot workspace component.
 * <p>
 * TODO: Make a version that extends this, like network panel case. Then
 * document in UML
 */
public class RasterPlotPanel extends JPanel {

    /**
     * Chart un-initialized instance.
     */
    private JFreeChart chart;

    /**
     * Initial size.
     */
    private static final Dimension PREFERRED_SIZE = new Dimension(500, 400);

    /**
     * Panel for chart.
     */
    private ChartPanel chartPanel = new ChartPanel(null);

    /**
     * Data model.
     */
    private RasterModel model;

    /**
     * Button panel.
     */
    private JPanel buttonPanel = new JPanel();

    /**
     * Construct a raster panel.
     *
     * @param rasterModel model underlying model
     */
    public RasterPlotPanel(final RasterModel rasterModel) {

        model = rasterModel;
        setPreferredSize(PREFERRED_SIZE);
        setLayout(new BorderLayout());

        addClearGraphDataButton();
        addPreferencesButton();

        add("Center", chartPanel);
        add("South", buttonPanel);

        init();
    }

    /**
     * Initialize Chart Panel.
     */
    public void init() {

        // Generate the graph
        chart = ChartFactory.createScatterPlot("", // Title
                "Iterations", // x-axis Label
                "Value(s)", // y-axis Label
                new XYSeriesCollection(model.getDataset()), // Dataset
                PlotOrientation.VERTICAL, // Plot Orientation
                true, // Show Legend
                true, // Use tooltips
                false // Configure chart to generate URLs?
        );
        XYItemRenderer renderer = ((XYPlot) chart.getPlot()).getRenderer();
        renderer.setSeriesPaint(0, Color.BLACK);
        double size = 1.0;
        double delta = size / 2.0;
        Shape shape1 = new Rectangle2D.Double(-delta, -delta, size, size);
        Shape shape2 = new Ellipse2D.Double(-delta, -delta, size, size);
        renderer.setSeriesShape(0, shape1);
        renderer.setSeriesShape(1, shape2);
        renderer.setSeriesShape(2, shape1);
        renderer.setSeriesShape(3, shape2);
        chartPanel.setChart(chart);
        chart.setBackgroundPaint(null);

        // // Create chart settings listener
        // model.addChartSettingsListener(new ChartSettingsListener() {
        //     public void chartSettingsUpdated(ChartModel theModel) {
        //
        //         // Handle range properties
        //         chart.getXYPlot().getRangeAxis().setAutoRange(model.isAutoRange());
        //         if (!model.isAutoRange()) {
        //             chart.getXYPlot().getRangeAxis().setRange(model.getRangeLowerBound(), model.getRangeUpperBound());
        //         }
        //
        //         // Handle domain properties
        //         if (model.isFixedWidth()) {
        //             chart.getXYPlot().getDomainAxis().setFixedAutoRange(model.getWindowSize());
        //         } else {
        //             chart.getXYPlot().getDomainAxis().setFixedAutoRange(-1);
        //             chart.getXYPlot().getDomainAxis().setAutoRange(true);
        //         }
        //     }
        // });
        //
        // // Invoke an initial event in order to set default settings
        // model.fireSettingsChanged();
    }

    /**
     * Remove all buttons from the button panel; used when customzing the
     * buttons on this panel.
     */
    public void removeAllButtonsFromToolBar() {
        buttonPanel.removeAll();
    }

    /**
     * Return button panel in case user would like to add custom buttons.
     *
     * @return
     */
    public JPanel getButtonPanel() {
        return buttonPanel;
    }

    /**
     * Add button for clearing graph data.
     */
    public void addClearGraphDataButton() {
        JButton clearButton = new JButton("Clear");
        clearButton.setAction(RasterPlotActions.getClearGraphAction(this));
        buttonPanel.add(clearButton);
    }

    /**
     * Add button for showing preferences.
     */
    public void addPreferencesButton() {
        JButton prefsButton = new JButton("Prefs");
        prefsButton.setHideActionText(true);
        prefsButton.setAction(RasterPlotActions.getPropertiesDialogAction(this));
        buttonPanel.add(prefsButton);
    }

    /**
     * Show properties dialog.
     */
    public void showPropertiesDialog() {
        ReflectivePropertyEditor editor = (new ReflectivePropertyEditor(model));
        JDialog dialog = editor.getDialog();
        dialog.setModal(true);
        dialog.pack();
        dialog.setLocationRelativeTo(null);
        dialog.setVisible(true);
    }

    /**
     * @return the chartPanel
     */
    public ChartPanel getChartPanel() {
        return chartPanel;
    }

    /**
     * @return the model
     */
    public RasterModel getRasterModel() {
        return model;
    }
}

File: src/main/kotlin/org/simbrain/plot/rasterchart/RasterPlotPanel.kt
/**
 * Swing panel for a [org.simbrain.plot.raster.RasterModel]: the JFreeChart scatter plot plus a
 * [ChartLegendPanel] legend strip and toolbar. The legend is Swing rather than JFreeChart's
 * in-chart title so each series can carry a hoverable remove control; series are deleted from
 * there, not from a toolbar button. Usable independently of the raster workspace component;
 * callers construct it and then call [init] to build the chart.
 */
package org.simbrain.plot.rasterchart

import org.jfree.chart.ChartFactory
import org.jfree.chart.ChartPanel
import org.jfree.chart.JFreeChart
import org.jfree.chart.axis.SymbolAxis
import org.jfree.chart.plot.PlotOrientation
import org.jfree.chart.renderer.AbstractRenderer
import org.jfree.chart.renderer.xy.XYItemRenderer
import org.simbrain.plot.ChartLegendPanel
import org.simbrain.plot.applySimbrainChartTheme
import org.simbrain.plot.raster.RasterModel
import org.simbrain.util.Theme
import org.simbrain.util.createEditorDialog
import org.simbrain.util.display
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Dimension
import java.awt.geom.Ellipse2D
import java.awt.geom.Rectangle2D
import javax.swing.JButton
import javax.swing.JPanel
import kotlin.math.max

class RasterPlotPanel(val rasterModel: RasterModel) : JPanel() {

    private lateinit var chart: JFreeChart

    val chartPanel: ChartPanel = ChartPanel(null)

    val buttonPanel: JPanel = JPanel()

    /**
     * Renderer object where things like dot color and size are set.
     */
    private lateinit var renderer: XYItemRenderer

    private val legendPanel = ChartLegendPanel()

    init {
        preferredSize = PREFERRED_SIZE
        layout = BorderLayout()

        addAddSeriesButton()
        addClearGraphDataButton()
        addPreferencesButton()

        add("Center", chartPanel)
        add("South", JPanel(BorderLayout()).apply {
            add(legendPanel, BorderLayout.NORTH)
            add(buttonPanel, BorderLayout.SOUTH)
        })
    }

    /**
     * Initialize Chart Panel.
     */
    fun init() {
        chart = ChartFactory.createScatterPlot(
            "",
            "Iterations",
            "Value(s)",
            rasterModel.dataset,
            PlotOrientation.VERTICAL,
            false,
            true,
            false
        )
        renderer = chart.xyPlot.renderer
        applyRowAxis()
        updateChartSettings()
        rasterModel.events.propertyChanged.on { updateChartSettings() }
        rasterModel.events.rasterConsumerAdded.on { rebuildLegend() }
        rasterModel.events.rasterConsumerRemoved.on { rebuildLegend() }
        chartPanel.chart = chart
        chart.applySimbrainChartTheme()
        rebuildLegend()
    }

    private fun rebuildLegend() {
        // Pin palette assignment to series order: lookupSeriesPaint hands out the next palette color
        // to whoever asks first, and swatches otherwise ask in reverse z-order during painting.
        (renderer as? AbstractRenderer)?.let { r ->
            rasterModel.rasterConsumerList.indices.forEach { r.lookupSeriesPaint(it) }
        }
        legendPanel.setEntries(rasterModel.rasterConsumerList.map { consumer ->
            ChartLegendPanel.Entry(
                label = rasterModel.dataset.getSeries(consumer.index).key.toString(),
                paint = { (renderer as? AbstractRenderer)?.lookupSeriesPaint(consumer.index) },
                onRemove = { rasterModel.removeDataSource(consumer) }
            )
        })
    }

    /**
     * Show the incoming array's component names down the row axis, e.g. neuron labels, falling back to the
     * row index for rows the producer did not name. The axis is rebuilt on every settings change, after
     * [applySimbrainChartTheme] has run, so it must theme itself or it comes up with dark-on-dark text and
     * opaque light grid bands under the dark theme.
     */
    private fun applyRowAxis() {
        val rows = max(max(rasterModel.rowCount, rasterModel.componentNames.size), 1)
        val labels = Array(rows) { row ->
            rasterModel.componentNames.getOrNull(row) ?: row.toString()
        }
        val axis = SymbolAxis("Value(s)", labels)
        axis.setRange(-0.5, rows - 0.5)
        axis.labelPaint = Theme.foreground
        axis.tickLabelPaint = Theme.foreground
        axis.axisLinePaint = Theme.divider
        axis.tickMarkPaint = Theme.divider
        axis.gridBandPaint = Theme.divider.let { Color(it.red, it.green, it.blue, 60) }
        axis.gridBandAlternatePaint = TRANSPARENT
        chart.xyPlot.rangeAxis = axis
    }

    fun updateChartSettings() {
        val size = rasterModel.dotSize.toDouble()
        val delta = size / 2.0
        val square = Rectangle2D.Double(-delta, -delta, size, size)
        val circle = Ellipse2D.Double(-delta, -delta, size, size)
        renderer.setSeriesShape(0, square)
        renderer.setSeriesShape(1, circle)
        renderer.setSeriesShape(2, square)
        renderer.setSeriesShape(3, circle)

        applyRowAxis()

        if (rasterModel.isFixedWidth) {
            chart.xyPlot.domainAxis.fixedAutoRange = rasterModel.windowSize.toDouble()
        } else {
            chart.xyPlot.domainAxis.fixedAutoRange = -1.0
            chart.xyPlot.domainAxis.isAutoRange = true
        }
    }

    /**
     * Remove all buttons from the button panel; used when customizing the buttons on this panel.
     */
    fun removeAllButtonsFromToolBar() {
        buttonPanel.removeAll()
    }

    fun addClearGraphDataButton() {
        val clearButton = JButton("Clear")
        clearButton.action = RasterPlotActions.getClearGraphAction(this)
        buttonPanel.add(clearButton)
    }

    fun addPreferencesButton() {
        val prefsButton = JButton("Prefs")
        prefsButton.hideActionText = true
        prefsButton.action = RasterPlotActions.getPropertiesDialogAction(this)
        buttonPanel.add(prefsButton)
    }

    fun addAddSeriesButton() {
        val addButton = JButton("Add")
        addButton.action = RasterPlotActions.getAddSourceAction(this)
        buttonPanel.add(addButton)
    }

    fun showPropertiesDialog() {
        rasterModel.createEditorDialog { updateChartSettings() }.display()
    }

    companion object {
        private val PREFERRED_SIZE = Dimension(500, 400)
        private val TRANSPARENT = Color(0, 0, 0, 0)
    }
}
